File: src/frankversnel/processing/component/ComponentManager.java
package frankversnel.processing.component;

import java.util.LinkedList;
import java.util.List;

public abstract class ComponentManager <T extends Component> {
	protected List<T> components = new LinkedList<T>();
	
	public void addComponent(T component) {
		components.add(component);
	}
	
	public abstract void processComponents();
	
	public boolean removeComponent(Component component) {
		return components.remove(component);
	}
	
}

File: src/main/scala/org/frankversnel/poortjes/common/component_manager.scala
package org.frankversnel.poortjes

import scala.collection.mutable.ListBuffer

import org.frankversnel.poortjes.util.DeltaTime

abstract class ComponentManager {
	type T

	private var components = ListBuffer[T]()

	def addComponent(component: Component) {
		forComponent(component) (components += _.asInstanceOf[T])
	}
	def removeComponent(component: Component) {
		forComponent(component) (components -= _.asInstanceOf[T])
	}

	protected def allComponents = components.toList

	def processComponents(update: Update): Unit
	protected def isCorrectType(component: Component): Boolean

	// Only processes the component if it is of type T
	private def forComponent[T: Manifest](c: Component) (onComponent: Component => Unit) {
		if(isCorrectType(c)) {
			onComponent(c)
		}
	}
}
